### PR TITLE
limine: reformat, update, build single arch

### LINKED ISSUE #304746: Build failure: limine
### Steps To Reproduce

1. `nix-build https://github.com/NixOS/nixpkgs/archive/master.tar.gz --attr limine` (behaviour is the same when trying to `nix-shell -p limine`)
2. Fails finding a riscv64 cross compiler toolchain.

### Build log
https://gist.github.com/T-O-R-U-S/0ed1610aeb91216fd688153d67799940

### Additional context

It asks for a RISC-V compiler because the package is configured with the `--enable-all` flag, so it fails to compile if any architecture is missing. Can be worked around by just enabling it for the platform that you are targeting, so long as that isn't RISC-V...

P.S: I only tried with the `--enable-x86_64` flag, and it seems to compile alright with that, but not with RISC-V.

### Notify maintainers

@48cf
@phip1611 

Thank you.

### Metadata

Please run `nix-shell -p nix-info --run "nix-info -m"` and paste the result.

```console
[user@system:~]$ nix-shell -p nix-info --run "nix-info -m"
 - system: `"aarch64-darwin"`
 - host os: `Darwin 23.0.0, macOS 14.0`
 - multi-user?: `yes`
 - sandbox: `no`
 - version: `nix-env (Nix) 2.21.1`
 - channels(root): `"nixpkgs"`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixpkgs`
```

---

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

### PR DESCRIPTION
**Pull request by surfaceflinger**
## Description of changes

Similar to what was done in #338953 

fixes #304746

I'd like someone to test building this on darwin in 4 variants:

- as is
- as is over pkgsCross
- allVariants = true
- allVariants = true over pkgsCross

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [x] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by alyssais** [comment]
This is fine I guess, but it'd still be good to understand why it broke if it used to work… 

I can try to bisect when I have spare capacity but I have a bit of a queue right now.

**Comment by phip1611** [comment]
Thanks for also working on this. I'm not sure

But please note that a useful Limine distribution needs a host-tool and **cross-compiled** bootloader-artifacts. 

So for example, on a x86 machine, one wants to be able to have:

- the x86 host tool 
- the EFI bootloader images for aarch64, i386, x86_64 and riscv


in a hypothetical scenario where one creates a OS in Rust that should be bootable for multiple platforms...

Sure, one can use pkgsCross and assemble something together but this will result in insane build times as all the cross toolchains need to be build. Right?

Therefore, it is an magnitude of order simpler to just use clang. As done in the CI of Limine: https://github.com/limine-bootloader/limine/blob/9b15ebed6213c56e9c6c5e99b65ee52d76077d4b/.github/workflows/check.yml#L22

Thanks for trying to fix this. But I'm very unsure how useful the Limine package will be with the proposed "host architecture only" approach.

**Comment by surfaceflinger** [comment]
If that's what you think then I'll try to get a flag for --enable-all done, however I don't think any bootloader in nixpkgs does that.
If an end user in the future does `boot.loader.limine.enable = true;` on their x86_64 machine, I don't think they want to have executables for loongarch64 lying around.

I would also like to get all limine deps using nix instead of downloading the prepackaged tarball but idk how it will go as I'm usually short on time now

**Comment by surfaceflinger** [comment]
also, I tested `pkgsCross` and it's not that bad imo, `pkgsi686Linux` and `aarch64-multiplatform` are available in binary caches (assuming you're on x86_64 linux), I think riscv64 is available too? I'm not sure rn. `loongarch64` toolchain definitely will have to be built from src but it's not that bad, at least on 5600x + pretty sure it's a common thing in Nix teams to have a shared binary cache filled by CI/CD.

**Comment by alyssais** [comment]
> in a hypothetical scenario where one creates a OS in Rust that should be bootable for multiple platforms...

In this scenario, wouldn't they need to be using pkgsCross anyway, to actually build their OS?

**Comment by phip1611** [comment]
>  that's what you think then I'll try to get a flag for --enable-all done

An overridable flag would be fine, as well.

> however I don't think any bootloader in nixpkgs does that.

Fair point! To be honest, when I added Limine to nixpkgs, I always had the "I'm an OS hacker and need a Multiboot2 bootloader that is cooler as GRUB" glasses on. For that, Limine is nice.

But Limine should also align with the other bootloaders in nixpkgs. Further, eventually, a boot.loader.limine.enable option would be cool.


TL;DR:

As long as a working `--enable-all` flag (attribute to the derivation) or the `symlinkJoin` approach mentioned above will work reasonably, I'm fine with that. I just don't want to limit end users // provide them with flexibility for using Limine as bootloader in their custom OS projects.

**Comment by alyssais** [comment]
pkgsCross is not generally intended to be used in Nixpkgs like that (there are some exceptions) — it's a layering violation and also slow, so the symlinkJoin approach would not be a good way to solve this.

An --enable-all package is fine if somebody can get it working — it seems like we're still talking around the fact that this used to work, doesn't any more, but nobody has checked why, but it's still unclear to me what makes limine special here, because as noted above, the user being able to avoid using pkgsCross just for limine doesn't really help much when they'll still need cross compilers for the rest of their OS.

**Comment by phip1611** [comment]
> it seems like we're still talking around the fact that this used to work, doesn't any more, but nobody has checked why, but

Agree.

**Comment by surfaceflinger** [comment]
@AsyaTheAbove lmk if you want to be added as a maintainer here in the same pr

**Comment by AsyaTheAbove** [comment]
@surfaceflinger yes that would be awesome. This is how the limine package must have been done since the beginning, LGTM. I've talked with upstream and they are aware that GCC can't build Limine for risc-v (no PIE support in GCC for riscv64-elf) but otherwise it would be awesome to be able to build with GCC as well.

**Comment by AsyaTheAbove** [comment]
Maybe, there could be easier options to enable individual architectures, as I did in my PR. My nix-based homebrew OS bootstrapper makes use of that. This would allow building for x86_32 without hostPlatform being x86_32, which is a legitimate use case. E.g. one could want to boot a homebrew kernel booted by x86_32 BIOS limine but switches to long mode itself, so one would want to target x86_64 for most packages but also be able to build x86_32 limine, and it would be much easier to build all-except-one targets.

**Comment by alyssais** [comment]
> E.g. one could want to boot a homebrew kernel booted by x86_32 BIOS limine but switches to long mode itself, so one would want to target x86_64 for most packages but also be able to build x86_32 limine, and it would be much easier to build all-except-one targets.

Is there enough in the pkgsi686Linux cache for it to be reasonable to get limine from there?

Having a list of targets to build (which could currently exclude RISC-V) sounds reasonable though.  It might be worth looking at the QEMU expression here for inspiration on how to handle target lists.

File: pkgs/by-name/li/limine/package.nix
Comment (phip1611) [inline pkgs/by-name/li/limine/package.nix:8]: nit: why not enableAll?

Comment (phip1611) [inline pkgs/by-name/li/limine/package.nix:8]: nit
```suggestion
  # The following options map to configure flags.
  allVariants ? false,
```

File: pkgs/by-name/li/limine/package.nix
Comment (phip1611) [inline pkgs/by-name/li/limine/package.nix:72]: is destdir needed/actually used?

File: pkgs/by-name/li/limine/package.nix
Comment (phip1611) [inline pkgs/by-name/li/limine/package.nix:97]: Probably yes

File: pkgs/by-name/li/limine/package.nix
Comment (phip1611) [inline pkgs/by-name/li/limine/package.nix:1]:
```suggestion
# Derivation containing the Limine host tool and the compiled bootloader
# artifacts for the configured architectures.
{
```

I personally prefer more verbosity, especially as this might not be clear to people out of the loop.

Comment (phip1611) [inline pkgs/by-name/li/limine/package.nix:1]:
```suggestion
# Derivation containing the Limine host tool and the compiled bootloader
# artifacts for the configured architectures.
{
```

I personally prefer more descriptive comments, especially as this might not be clear to people out of the loop.

**Comment by emilazy** [comment]
FWIW, I have some hopes that we’ll be able to switch new legacy BIOS installations of NixOS to use Limine instead of GRUB at some point, given its much better maintenance state and the problems GRUB has caused us in the past. So I’m definitely in favour of aligning this with other bootloaders in Nixpkgs by default rather than focusing on the OS developer use‐case.

(There are no concrete plans for this and I have not coordinated it with anyone beyond bringing the idea up a couple times, so don’t take this as any official statement. But the problems with GRUB are well‐known and I feel a BIOS bootloader with decent maintenance would be an easier sell than the U‐Boot UEFI shim that has been proposed to solve the GRUB problem in the past.)

I will test the build on Darwin.

**Comment by phip1611** [comment]
> focusing on the OS developer use‐case.

As long as there are configurable options, that is perfectly fine and valid. I absolutely see and also support your point.

**Comment by emilazy** [comment]
`allVariants` is broken on `x86_64-darwin`:

```
limine> ld.lld \
limine> 	-T'/nix/tmp/nix-build-limine-8.0.11.drv-0/limine-8.0.11/common-uefi-loongarch64/linker_nomap.ld' \
limine> 	'/nix/tmp/nix-build-limine-8.0.11.drv-0/limine-8.0.11/common-uefi-loongarch64/limine-efi/src/crt0-efi-loongarch64.S.o' '/nix/tmp/nix-build-limine-8.0.11.drv-0/limine-8.0.11/common-uefi-loongarch64/limine-efi/src/reloc_loongarch64.c.o' '/nix/tmp/nix-build-limine-8.0.11.drv-0/limine-8.0.11/common-uefi-loongarch64/./cc-runtime.s2.o' '/nix/tmp/nix-build-limine-8.0.11.drv-0/limine-8.0.11/common-uefi-loongarch64/./crypt/blake2b.o' '/nix/tmp/nix-build-limine-8.0.11.drv-0/limine-8.0.11/common-uefi-loongarch64/./drivers/disk.s2.o' '/nix/tmp/nix-build-limine-8.0.11.drv-0/limine-8.0.11/common-uefi-loongarch64/./drivers/edid.o' '/nix/tmp/nix-build-limine-8.0.11.drv-0/limine-8.0.11/common-uefi-loongarch64/./drivers/gop.o' '/nix/tmp/nix-build-limine-8.0.11.drv-0/limine-8.0.11/common-uefi-loongarch64/./drivers/serial.o' '/nix/tmp/nix-build-limine-8.0.11.drv-0/limine-8.0.11/common-uefi-loongarch64/./drivers/vbe.o' '/nix/tmp/nix-build-limine-8.0.11.drv-0/limine-8.0.11/common-uefi-loongarch64/./drivers/vga_textmode.o' '/nix/tmp/nix-build-limine-8.0.11.drv-0/limine-8.0.11/common-uefi-loongarch64/./entry.s2.o' '/nix/tmp/nix-build-limine-8.0.11.drv-0/limine-8.0.11/common-uefi-loongarch64/./entry.s3.o' '/nix/tmp/nix-build-limine-8.0.11.drv-0/limine-8.0.11/common-uefi-loongarch64/./flanterm/backends/fb.o' '/nix/tmp/nix-build-limine-8.0.11.drv-0/limine-8.0.11/common-uefi-loongarch64/./flanterm/flanterm.o' '/nix/tmp/nix-build-limine-8.0.11.drv-0/limine-8.0.11/common-uefi-loongarch64/./fs/ext2.s2.o' '/nix/tmp/nix-build-limine-8.0.11.drv-0/limine-8.0.11/common-uefi-loongarch64/./fs/fat32.s2.o' '/nix/tmp/nix-build-limine-8.0.11.drv-0/limine-8.0.11/common-uefi-loongarch64/./fs/file.s2.o' '/nix/tmp/nix-build-limine-8.0.11.drv-0/limine-8.0.11/common-uefi-loongarch64/./fs/iso9660.s2.o' '/nix/tmp/nix-build-limine-8.0.11.drv-0/limine-8.0.11/common-uefi-loongarch64/./lib/acpi.o' '/nix/tmp/nix-build-limine-8.0.11.drv-0/limine-8.0.11/common-uefi-loongarch64/./lib/config.o' '/nix/tmp/nix-build-limine-8.0.11.drv-0/limine-8.0.11/common-uefi-loongarch64/./lib/elf.o' '/nix/tmp/nix-build-limine-8.0.11.drv-0/limine-8.0.11/common-uefi-loongarch64/./lib/elsewhere.o' '/nix/tmp/nix-build-limine-8.0.11.drv-0/limine-8.0.11/common-uefi-loongarch64/./lib/fb.o' '/nix/tmp/nix-build-limine-8.0.11.drv-0/limine-8.0.11/common-uefi-loongarch64/./lib/fdt.o' '/nix/tmp/nix-build-limine-8.0.11.drv-0/limine-8.0.11/common-uefi-loongarch64/./lib/getchar.o' '/nix/tmp/nix-build-limine-8.0.11.drv-0/limine-8.0.11/common-uefi-loongarch64/./lib/gterm.o' '/nix/tmp/nix-build-limine-8.0.11.drv-0/limine-8.0.11/common-uefi-loongarch64/./lib/guid.o' '/nix/tmp/nix-build-limine-8.0.11.drv-0/limine-8.0.11/common-uefi-loongarch64/./lib/image.o' '/nix/tmp/nix-build-limine-8.0.11.drv-0/limine-8.0.11/common-uefi-loongarch64/./lib/libc.o' '/nix/tmp/nix-build-limine-8.0.11.drv-0/limine-8.0.11/common-uefi-loongarch64/./lib/libc.s2.o' '/nix/tmp/nix-build-limine-8.0.11.drv-0/limine-8.0.11/common-uefi-loongarch64/./lib/memory.s2.o' '/nix/tmp/nix-build-limine-8.0.11.drv-0/limine-8.0.11/common-uefi-loongarch64/./lib/misc.o' '/nix/tmp/nix-build-limine-8.0.11.drv-0/limine-8.0.11/common-uefi-loongarch64/./lib/misc.s2.o' '/nix/tmp/nix-build-limine-8.0.11.drv-0/limine-8.0.11/common-uefi-loongarch64/./lib/panic.s2.o' '/nix/tmp/nix-build-limine-8.0.11.drv-0/limine-8.0.11/common-uefi-loongarch64/./lib/part.s2.o' '/nix/tmp/nix-build-limine-8.0.11.drv-0/limine-8.0.11/common-uefi-loongarch64/./lib/print.s2.o' '/nix/tmp/nix-build-limine-8.0.11.drv-0/limine-8.0.11/common-uefi-loongarch64/./lib/rand.o' '/nix/tmp/nix-build-limine-8.0.11.drv-0/limine-8.0.11/common-uefi-loongarch64/./lib/stb_image.o' '/nix/tmp/nix-build-limine-8.0.11.drv-0/limine-8.0.11/common-uefi-loongarch64/./lib/term.o' '/nix/tmp/nix-build-limine-8.0.11.drv-0/limine-8.0.11/common-uefi-loongarch64/./lib/time.o' '/nix/tmp/nix-build-limine-8.0.11.drv-0/limine-8.0.11/common-uefi-loongarch64/./lib/trace.s2.o' '/nix/tmp/nix-build-limine-8.0.11.drv-0/limine-8.0.11/common-uefi-loongarch64/./lib/uri.o' '/nix/tmp/nix-build-limine-8.0.11.drv-0/limine-8.0.11/common-uefi-loongarch64/./libfdt/fdt.o' '/nix/tmp/nix-build-limine-8.0.11.drv-0/limine-8.0.11/common-uefi-loongarch64/./libfdt/fdt_addresses.o' '/nix/tmp/nix-build-limine-8.0.11.drv-0/limine-8.0.11/common-uefi-loongarch64/./libfdt/fdt_check.o' '/nix/tmp/nix-build-limine-8.0.11.drv-0/limine-8.0.11/common-uefi-loongarch64/./libfdt/fdt_empty_tree.o' '/nix/tmp/nix-build-limine-8.0.11.drv-0/limine-8.0.11/common-uefi-loongarch64/./libfdt/fdt_overlay.o' '/nix/tmp/nix-build-limine-8.0.11.drv-0/limine-8.0.11/common-uefi-loongarch64/./libfdt/fdt_ro.o' '/nix/tmp/nix-build-limine-8.0.11.drv-0/limine-8.0.11/common-uefi-loongarch64/./libfdt/fdt_rw.o' '/nix/tmp/nix-build-limine-8.0.11.drv-0/limine-8.0.11/common-uefi-loongarch64/./libfdt/fdt_strerror.o' '/nix/tmp/nix-build-limine-8.0.11.drv-0/limine-8.0.11/common-uefi-loongarch64/./libfdt/fdt_sw.o' '/nix/tmp/nix-build-limine-8.0.11.drv-0/limine-8.0.11/common-uefi-loongarch64/./libfdt/fdt_wip.o' '/nix/tmp/nix-build-limine-8.0.11.drv-0/limine-8.0.11/common-uefi-loongarch64/./menu.o' '/nix/tmp/nix-build-limine-8.0.11.drv-0/limine-8.0.11/common-uefi-loongarch64/./mm/mtrr.o' '/nix/tmp/nix-build-limine-8.0.11.drv-0/limine-8.0.11/common-uefi-loongarch64/./mm/pmm.o' '/nix/tmp/nix-build-limine-8.0.11.drv-0/limine-8.0.11/common-uefi-loongarch64/./mm/pmm.s2.o' '/nix/tmp/nix-build-limine-8.0.11.drv-0/limine-8.0.11/common-uefi-loongarch64/./mm/vmm.o' '/nix/tmp/nix-build-limine-8.0.11.drv-0/limine-8.0.11/common-uefi-loongarch64/./protos/chainload.o' '/nix/tmp/nix-build-limine-8.0.11.drv-0/limine-8.0.11/common-uefi-loongarch64/./protos/chainload_next.o' '/nix/tmp/nix-build-limine-8.0.11.drv-0/limine-8.0.11/common-uefi-loongarch64/./protos/limine.o' '/nix/tmp/nix-build-limine-8.0.11.drv-0/limine-8.0.11/common-uefi-loongarch64/./protos/linux.o' '/nix/tmp/nix-build-limine-8.0.11.drv-0/limine-8.0.11/common-uefi-loongarch64/./protos/linux_risc.o' '/nix/tmp/nix-build-limine-8.0.11.drv-0/limine-8.0.11/common-uefi-loongarch64/./protos/multiboot1.o' '/nix/tmp/nix-build-limine-8.0.11.drv-0/limine-8.0.11/common-uefi-loongarch64/./protos/multiboot2.o' '/nix/tmp/nix-build-limine-8.0.11.drv-0/limine-8.0.11/common-uefi-loongarch64/./pxe/pxe.s2.o' '/nix/tmp/nix-build-limine-8.0.11.drv-0/limine-8.0.11/common-uefi-loongarch64/./pxe/tftp.s2.o' '/nix/tmp/nix-build-limine-8.0.11.drv-0/limine-8.0.11/common-uefi-loongarch64/./sys/a20.s2.o' '/nix/tmp/nix-build-limine-8.0.11.drv-0/limine-8.0.11/common-uefi-loongarch64/./sys/cpu_riscv.o' '/nix/tmp/nix-build-limine-8.0.11.drv-0/limine-8.0.11/common-uefi-loongarch64/./sys/e820.s2.o' '/nix/tmp/nix-build-limine-8.0.11.drv-0/limine-8.0.11/common-uefi-loongarch64/./sys/exceptions.s2.o' '/nix/tmp/nix-build-limine-8.0.11.drv-0/limine-8.0.11/common-uefi-loongarch64/./sys/gdt.s2.o' '/nix/tmp/nix-build-limine-8.0.11.drv-0/limine-8.0.11/common-uefi-loongarch64/./sys/idt.o' '/nix/tmp/nix-build-limine-8.0.11.drv-0/limine-8.0.11/common-uefi-loongarch64/./sys/idt.s2.o' '/nix/tmp/nix-build-limine-8.0.11.drv-0/limine-8.0.11/common-uefi-loongarch64/./sys/lapic.o' '/nix/tmp/nix-build-limine-8.0.11.drv-0/limine-8.0.11/common-uefi-loongarch64/./sys/pic.o' '/nix/tmp/nix-build-limine-8.0.11.drv-0/limine-8.0.11/common-uefi-loongarch64/./sys/smp.o' '/nix/tmp/nix-build-limine-8.0.11.drv-0/limine-8.0.11/common-uefi-loongarch64/./lib/spinup.o' '/nix/tmp/nix-build-limine-8.0.11.drv-0/limine-8.0.11/common-uefi-loongarch64/./menu_thunk.o' '/nix/tmp/nix-build-limine-8.0.11.drv-0/limine-8.0.11/common-uefi-loongarch64/./efi_thunk.o' -nostdlib -z max-page-size=0x1000 -gc-sections -m elf64loongarch --no-relax -pie -z text -o '/nix/tmp/nix-build-limine-8.0.11.drv-0/limine-8.0.11/common-uefi-loongarch64/limine_nomap.elf'
limine> ld.lld: error: unknown emulation: elf64loongarch
```

**Comment by AsyaTheAbove** [comment]
I'm closing #338953 in favor of this

> Is there enough in the pkgsi686Linux cache for it to be reasonable to get limine from there?

Not sure
> Having a list of targets to build (which could currently exclude RISC-V) sounds reasonable though. It might be worth looking at the QEMU expression here for inspiration on how to handle target lists.

IIRC, only GCC has issues with RISC-V, which is an issue with GCC itself, so it would be best to keep RISC-V in the list of targets to build when building with Clang
> focusing on the OS developer use‐case.

I agree with @phip1611 on this, it is entirely possible to support both, and in fact the use-cases aren't that much different other than a few QoL changes for the OSdev use-case.

**Comment by emilazy** [comment]
No objection to supporting both. Just saying that I think the default should be a configuration we’d want to ship as a NixOS bootloader.

**Comment by surfaceflinger** [comment]
@emilazy which LLVM version is used on Darwin in `llvmPackages`? https://github.com/NixOS/nixpkgs/blob/261a20d9080cfc1bd7730c6ab1e1d6491d68cf15/pkgs/top-level/all-packages.nix#L15439-L15456 
suggests that darwin uses 16 while `elf64loongarch` was added here https://github.com/llvm/llvm-project/commit/6084ee742064cf8121444febee342133b0cb7abd and this commit is available since llvm 17.

I guess it should be enough to switch to `llvmPackages_latest.stdenv`

File: pkgs/by-name/li/limine/package.nix
Comment (surfaceflinger) [reply pkgs/by-name/li/limine/package.nix]: nope

**Comment by surfaceflinger** [comment]
I'll try later to do targets similar to how it's done in QEMU derivation but for now it's 1am for me

**Comment by emilazy** [comment]
Yeah, sounds reasonable. We’re going to bump to 19 on ~all platforms after 24.11 is branched off, at which point we should unpin this, but I would suggest using `clang18Stdenv` for now. `llvmPackages_latest` won’t work as we don’t use aliases in‐tree. (Not sure what the difference between `clang18Stdenv` and `llvmPackages_18.stdenv` is…)

File: pkgs/by-name/li/limine/package.nix
Comment (AsyaTheAbove) [inline pkgs/by-name/li/limine/package.nix:67]: I believe NASM is only needed when building for x86

**Comment by surfaceflinger** [comment]
this one should be a banger unless if i messed up the logic somewhere or it can be done simpler somehow

**Review by phip1611 — APPROVED** [review]

**Comment by surfaceflinger** [comment]
hm, not sure why ofborg is failing, it builds correctly locally..

File: pkgs/by-name/li/limine/package.nix
Comment (phip1611) [inline pkgs/by-name/li/limine/package.nix:86]: The order is relevant. Now, `dev` is the default attribute.

```suggestion
  outputs = [
    out
```

**Review by phip1611 — CHANGES_REQUESTED** [review]
Now, `limine.dev` is the default output due to the re-ordering but it should be `limine.out`. `limine.dev` only contains the header file `limine.h`.

**Comment by surfaceflinger** [comment]
> It's because clang18Stdenv is an alias. Use the stdenv attribute of the llvmPackages you're using.

hmm, okay, I switched to clang18Stdenv because of https://github.com/NixOS/nixpkgs/pull/341087#issuecomment-2357120817
i'll do llvmPackages_18.stdenv instead

**Comment by phip1611** [comment]
I can confirm that `limineX = limine.override ({ enableAll = true; });` is sufficient for my personal use-cases. But I don't know about the MacOS situation, unfortunately.

**Comment by emilazy** [comment]
Sorry about the `clang*Stdenv` thing, I don’t know as much about our LLVM packaging as I should. Trust Alyssa over me :)

**Review by phip1611 — APPROVED** [review]

**Comment by surfaceflinger** [comment]
i'll have to setup a vm to fix that, unless someone wants to check that and send me a patch

**Comment by emilazy** [comment]
I can try to poke around if you have any idea at all about what the issue might be related to.

**Comment by surfaceflinger** [comment]
try planting `set -x` here https://github.com/osdev0/freestanding-toolchain/blob/trunk/freestanding-toolchain#L50-L68 by generating a patch from the src tarball, it should contain freestanding-toolchain. this thing tests the toolchain if found and if it fails then it looks for gcc.

also, freestanding-toolchain output might be redirected to a file `toolchain-files/bios-i686-toolchain.mk` so `nix develop` + genericBuild will be useful

you can also try forcing llvm with `TOOLCHAIN_FOR_TARGET=llvm`

**Review by lzcunt — APPROVED** [review]

**Comment by emilazy** [comment]
Sorry for not getting back to this yet. It should be fine to merge without Darwin support, but I’m going to throw in a ping to @NixOS/darwin-maintainers in case anyone can figure out the problem here quickly.

**Comment by surfaceflinger** [comment]
ok then, I'll readd `badPlatforms = platforms.darwin;`

**Review by lzcunt — APPROVED** [review]

**Review by emilazy — APPROVED** [review]

**Comment by pxseu** [comment]
@emilazy configure runs `clang -target i686-unknown-none -x c -c - -o /dev/null`
and it errors out with `limine> clang: error: unsupported option '-arch' for target 'i686-unknown-none'`

I don't have time rn to dig deeper but I think wrapper for clang is adding some colliding args?

**Comment by alyssais** [comment]
Huh, that shouldn't happen since https://github.com/NixOS/nixpkgs/commit/12b0e8ac74f82b585d43803bac89df08dca5285e.

**Pull request merged**
